clippy_lints/src/int_plus_one.rs

```diff
@@ -0,0 +1,183 @@
+//! lint on blocks unnecessarily using >= with a + 1 or - 1
+
+use rustc::lint::*;
+use syntax::ast::*;
+
+use utils::{span_lint_and_then, snippet_opt};
+
+/// **What it does:** Checks for usage of `x >= y + 1` or `x - 1 >= y` (and `<=`) in a block
+///
+///
+/// **Why is this bad?** Readability -- better to use `> y` instead of `>= y + 1`.
+///
+/// **Known problems:** None.
+///
+/// **Example:**
+/// ```rust
+/// x >= y + 1
+/// ```
+///
+/// Could be written:
+///
+/// ```rust
+/// x > y
+/// ```
+declare_lint! {
+    pub INT_PLUS_ONE,
+    Allow,
+    "instead of using x >= y + 1, use x > y"
+}
+
+pub struct IntPlusOne;
+
+impl LintPass for IntPlusOne {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(INT_PLUS_ONE)
+    }
+}
+
+// cases:
+// BinOpKind::Ge
+// x >= y + 1
+// x - 1 >= y
+//
+// BinOpKind::Le
+// x + 1 <= y
+// x <= y - 1
+
+enum Side {
+    LHS,
+    RHS,
+}
+
+impl IntPlusOne {
+    #[allow(cast_sign_loss)]
+    fn check_lit(&self, lit: &Lit, target_value: i128) -> bool {
+        if let LitKind::Int(value, ..) = lit.node {
+            return value == (target_value as u128)
+        }
+        false
+    }
+
+    fn check_binop(&self, cx: &EarlyContext, block: &Expr, binop: BinOpKind, lhs: &Expr, rhs: &Expr) -> Option<(bool, Option<String>)> {
```

Comment:
I'm very confused as to how this function successfully compiles ( the return type matches no returned value)

Reply:
Good catch -- I started moving toward returning the (bool, String) tuple I described above, but I forgot to remove the return value.

I'm not sure why the compiler didn't catch that though. AFAIK there is no implicit None returned.